From 32efbeb52d15ffa5b7dc60f9d1f8a2f6cb13ecb9 Mon Sep 17 00:00:00 2001
From: Kazunori Kobayashi <kkobayas@igel.co.jp>
Date: Mon, 13 Aug 2012 19:02:47 +0900
Subject: [PATCH 47/97] dfb-video-example: break out of event loop when
 playback position runs past the total duration

An EOS event may not be caught when trying to seek past the end of the stream
because GStreamer cancels the EOS event while flush processing is being
performed.
This patch resolves this problem by breaking out of the event loop if the
playback position goes beyond the total stream duration.
---
 ext/directfb/dfb-video-example.c |   10 ++++++++++
 1 file changed, 10 insertions(+)

diff --git a/ext/directfb/dfb-video-example.c b/ext/directfb/dfb-video-example.c
index b7f8793..18ab707 100644
--- a/ext/directfb/dfb-video-example.c
+++ b/ext/directfb/dfb-video-example.c
@@ -288,6 +288,9 @@ event_loop (GstElement * pipeline)
         s = gst_message_get_structure (message);
 
         if (gst_structure_has_name (s, "FrameRendered")) {
+          GstFormat fmt = GST_FORMAT_TIME;
+          gint64 total_duration;
+
           if (playback_rate == 1.0)
             break;
 
@@ -296,6 +299,13 @@ event_loop (GstElement * pipeline)
           memcpy (&prev_tv, &cur_tv, sizeof (prev_tv));
 
           position += playback_rate * duration;
+          gst_element_query_duration (pipeline, &fmt, &total_duration);
+
+          if (position < 0 || position > total_duration) {
+            /* finish when going beyond playback time */
+            loop = FALSE;
+            break;
+          }
 
           if (!gst_element_seek (pipeline, 1.0, GST_FORMAT_TIME,
                   GST_SEEK_FLAG_FLUSH | GST_SEEK_FLAG_ACCURATE,
-- 
1.7.10.4

